### PR TITLE
[SPARK-25417][SQL] ArrayContains function may return incorrect result when right expression is implicitly down casted

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
In ArrayContains, we currently cast the right hand side expression to match the element type of the left hand side Array. This may result in down casting and may return wrong result or questionable result.

Example :
```SQL
spark-sql> select array_contains(array(1), 1.34);
true
```
```SQL
spark-sql> select array_contains(array(1), 'foo');
null
```

We should safely coerce both left and right hand side expressions.
## How was this patch tested?
Added tests in DataFrameFunctionsSuite

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -1331,26 +1331,31 @@ case class ArrayContains(left: Expression, right: Expression)
...
+        TypeCheckResult.TypeCheckFailure("Null typed values cannot be used as arguments")
+      case (ArrayType(e1, _), e2) if e1.sameType(e2) =>
+        TypeUtils.checkForOrderingExpr(right.dataType, s"function $prettyName")
+      case _ => TypeCheckResult.TypeCheckFailure(s"Input to function $prettyName should have " +
+        s"been ${ArrayType.simpleString} followed by a value with same element type, but it's " +
+        s"[${left.dataType.catalogString}, ${right.dataType.catalogString}].")
     }
   }
 
+
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: can you revert this change?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 2]
```diff
@@ -735,6 +735,44 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
...
+      Seq(Row(true), Row(true))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(array(1)), array(1.23D))"),
+      Seq(Row(false), Row(false))
+    )
+
+    intercept[AnalysisException] {
+      df.selectExpr("array_contains(array(1), 1.23)")
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Why does this fail?

Comment (ueshin) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Also could you check the error message as well?

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @ueshin This fails since 1.23 is parsed as decimal(3, 2). And `findTightestCommonType` can not find a common type between integer and decimal(3,2).  I specifically added this case to draw it to your attention to see if we are okay with this behaviour :-). If we had it as `1.23D` then we would have been able to find a common type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 1 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @ueshin Sure.. sorry.. didn't notice it.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @ueshin Did you want me to check the error message for all the cases ? or just a couple of them ?

Comment (ueshin) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: As for the behavior, we might be better to use `findWiderTypeWithoutStringPromotion` instead of `findTightestCommonType`.
And as for checking the error message, just those you added.

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @ueshin Thank you. I also thought of using `findWiderTypeWithoutStringPromotion` but later changed it to `findTightestCommonType`. One question i had was, can `findWiderTypeWithoutStringPromotion` do a lossy conversion. From the class description it seems only findTightestCommonType does a absolute safe casting ? Please let me know what you think ..

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @ueshin FYI - i just pushed the changes addressing your comments. I have kept `findTightestCommonType` as is for now. I will change it based on your input and adjust the tests accordingly.

**Comment by dilipbiswal** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 2 above)
Comment (ueshin) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Good point. Yes, I can do a lossy conversion.
Seems like `BinaryComparison` uses wider DecimalType, so we could follow the behavior to use `findWiderTypeWithoutStringPromotion`.
cc @gatorsmile @cloud-fan

**Comment by maropu** [comment]
plz update the migration guide.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 2 above)
Comment (ueshin) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Hmm, if we can `array_contains(array(1), '1')` in 2.3 as https://github.com/apache/spark/pull/22408#issuecomment-421223501, we should use `findWiderCommonType`?

**Comment by maropu** [comment]
How about this decimal case?
```
// v2.3.1
scala> spark.range(10).selectExpr("cast(id AS decimal(9, 0)) as value").selectExpr("array_contains(array(1, 2, 3), value)").show
+--------------------------------------------------+
|array_contains(array(1, 2, 3), CAST(value AS INT))|
+--------------------------------------------------+
|                                             false|
|                                              true|
|                                              true|
|                                              true|
|                                             false|
|                                             false|
|                                             false|
|                                             false|
|                                             false|
|                                             false|
+--------------------------------------------------+

// this patch
scala> spark.range(10).selectExpr("cast(id AS decimal(9, 0)) as value").selectExpr("array_contains(array(1, 2, 3), value)").show
org.apache.spark.sql.AnalysisException: cannot resolve 'array_contains(array(1, 2, 3), `value`)' due to data type mismatch: Input to function array_contains should have been array followed by a value with same element type, but it's [array<int>, decimal(9,0)].; line 1 pos 0;
'Project [unresolvedalias(array_contains(array(1, 2, 3), value#2), Some(<function1>))]
+- Project [cast(id#0L as decimal(9,0)) AS value#2]
   +- Range (0, 10, step=1, splits=Some(4))
```

**Comment by dilipbiswal** [comment]
@maropu This is the case we were discussing for which @ueshin suggested using `findWiderTypeWithoutStringPromotion`. Lets see what @cloud-fan and @gatorsmile think and we will do accordingly. We are picking a much restrictive cast (since we are injecting this implicitly).

**Comment by gatorsmile** [comment]
What is the corresponding behavior in Presto?

**Comment by gatorsmile** [comment]
My general idea is to avoid risky implicit type casting at the beginning. We can relax it in the future, if needed. After all, users can manually cast the types after seeing the reasonable error message. This should not be a big deal. However, returning a confusing result due to implicit type casting is not good in general.

**Comment by dilipbiswal** [comment]
@ushin @gatorsmile Here are the results from presto. Please let me know if you want me try any case in particular. One thing to note is that presto allows comparison between int and decimal. In our `findTightestCommonType` we don't do the promotion.

``` SQL
presto:default> select contains(array[1,2,3], '1');
Query 20180914_053612_00006_pru6h failed: line 1:8: Unexpected parameters (array(integer), varchar(1)) for function contains. Expected: contains(array(T), T) T:comparable
select contains(array[1,2,3], '1')

presto:default> select contains(array[1,2,3], 'foo');
Query 20180914_053729_00007_pru6h failed: line 1:8: Unexpected parameters (array(integer), varchar(3)) for function contains. Expected: contains(array(T), T) T:comparable
select contains(array[1,2,3], 'foo')

presto:default> select contains(array['1','2','3'], 1);
Query 20180914_053850_00008_pru6h failed: line 1:8: Unexpected parameters (array(varchar(1)), integer) for function contains. Expected: contains(array(T), T) T:comparable
select contains(array['1','2','3'], 1)

presto:default> select contains(array[1,2,3], cast(1.0 as decimal(10,2)));
 _col0 
-------
 true  
(1 row)

presto:default> select contains(array[1,2,3], cast(1.0 as double));
 _col0 
-------
 true  
(1 row)
```

**Comment by cloud-fan** [comment]
what does presto return for `select array_contains(array(1), 1.34);`?

**Comment by dilipbiswal** [comment]
@cloud-fan It turns false.
```
presto:default> select contains(array[1], 1.34)
             -> ;
 _col0 
-------
 false 
(1 row)
```

**Comment by cloud-fan** [comment]
LGTM. I think the last piece is the migration guide, to explain what changed from 2.3 to 2.4

**Comment by dilipbiswal** [comment]
@cloud-fan Thanks a lot. Let me take a stab at the migration guide changes. 
cc @ueshin

**Comment by dilipbiswal** [comment]
@ueshin @cloud-fan @gatorsmile Can you please look at the doc changes when you get a chance.. Thanks.

File: docs/sql-programming-guide.md [hunk 3]
```diff
@@ -1879,6 +1879,80 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
+        </tr>
+        <tr>
+          <th>
+            <b>SELECT <br> array_contains(array(1), 1.34);</b>
+          </th>
+          <th>
+            <b>true</b>
+          </th>
+          <th>
+            <b>AnalysisException is thrown since integer type can not be promoted to decimal type in a loss-less manner.</b>
```
Comment (cloud-fan) [inline docs/sql-programming-guide.md]: hhmm, I think we can convert int to `DecimalType(10, 0)` without losing precision?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 4]
```diff
@@ -1331,23 +1331,27 @@ case class ArrayContains(left: Expression, right: Expression)
...
 
-  override def inputTypes: Seq[AbstractDataType] = right.dataType match {
-    case NullType => Seq.empty
-    case _ => left.dataType match {
-      case n @ ArrayType(element, _) => Seq(n, element)
+  override def inputTypes: Seq[AbstractDataType] = {
+    (left.dataType, right.dataType) match {
+      case (_, NullType) => Seq.empty
+      case (ArrayType(e1, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(e1, e2) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: I think we have a bug in the `findTightestCommonType`. For an int and a decimal, `findTightestCommonType` can return a value if the decimal's precision is bigger than int. But it can't return a value if the decimal's precision is small.

File: docs/sql-programming-guide.md (hunk 3 above)
Comment (dilipbiswal) [reply docs/sql-programming-guide.md]: @cloud-fan Yes..Actually, i think, the logic to handle int and decimal promotion can be improved. I like presto's behaviour here -
int, decimal(3, 2) => decimal(12, 2)
int, decimal(4, 3) => decimal(13, 3)
int, decimal(11, 3) => decimal(14, 3)

Comment (dilipbiswal) [reply docs/sql-programming-guide.md]: @cloud-fan I have taken a stab at fixing this at https://github.com/dilipbiswal/spark/tree/find_tightest
May i request you to please take a look ? I have run all the tests and they look ok. If you are okay, i will open a PR for this and once that gets in, we can look at this PR.

**Comment by cloud-fan** [comment]
Since it's a little risky to backport https://github.com/apache/spark/pull/22448 , I'd like to merge this PR without https://github.com/apache/spark/pull/22448 . Can you fix the conflict? thanks!

File: docs/sql-programming-guide.md [hunk 5]
```diff
@@ -1879,6 +1879,80 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
+          <th>
+            <b>false</b>
+          </th>
+          <th>
+            <b>In Spark 2.4, both left and right parameters are  promoted to array(double) and double type respectively.</b>
+          </th>
+        </tr>
+        <tr>
+          <th>
+            <b>SELECT <br> array_contains(array(1), 1.34);</b>
```
Comment (cloud-fan) [inline docs/sql-programming-guide.md]: can we remove this example? or pick a different one that can work. Basically it should work if we fix the bug with https://github.com/apache/spark/pull/22448

Comment (dilipbiswal) [reply docs/sql-programming-guide.md]: @cloud-fan OK.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 6]
```diff
@@ -735,6 +735,60 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
...
+      Seq(Row(true), Row(true))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(array(1)), array(1.23D))"),
+      Seq(Row(false), Row(false))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(array(1)), array(1.23))"),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: hmm? shouldn't this fail because of the bug?

File: docs/sql-programming-guide.md [hunk 7]
```diff
@@ -1879,6 +1879,66 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
+        </tr>
+        <tr>
+          <th>
+            <b>SELECT <br> array_contains(array(1), '1');</b>
+          </th>
+          <th>
+            <b>true</b>
+          </th>
+          <th>
+            <b>AnalysisException is thrown since integer type can not be promoted to string type in a loss-less manner.</b>
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply docs/sql-programming-guide.md]: Ah then it's fine, we don't need to change anything here.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 6 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @cloud-fan Yes. it should :-) I think i had changed this test case to verify the fix to tighestCommonType.. and pushed it by mistake. Sorry about it.

**Comment by cloud-fan** [comment]
LGTM

**Comment by dilipbiswal** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4!
